homeassistant/components/light/group.py

```diff
@@ -0,0 +1,280 @@
+        def async_state_changed_listener(entity_id: str, old_state: State,
+                                         new_state: State):
+            """Handle child updates."""
+            self.async_schedule_update_ha_state(True)
+
+        self._async_unsub_state_changed = async_track_state_change(
+            self.hass, self._entity_ids, async_state_changed_listener)
+
+    async def async_will_remove_from_hass(self):
+        """Callback when removed from HASS."""
+        if self._async_unsub_state_changed:
+            self._async_unsub_state_changed()
+            self._async_unsub_state_changed = None
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def state(self) -> str:
+        """Return the state."""
+        return self._state
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self._state == STATE_ON
+
+    @property
+    def brightness(self) -> Optional[int]:
+        """Return the brightness of this light between 0..255."""
+        return self._brightness
+
+    @property
+    def xy_color(self) -> Optional[Tuple[float, float]]:
+        """Return the XY color value [float, float]."""
+        return self._xy_color
+
+    @property
+    def rgb_color(self) -> Optional[Tuple[int, int, int]]:
+        """Return the RGB color value [int, int, int]."""
+        return self._rgb_color
+
+    @property
+    def color_temp(self) -> Optional[int]:
+        """Return the CT color value in mireds."""
+        return self._color_temp
+
+    @property
+    def min_mireds(self) -> Optional[int]:
+        """Return the coldest color_temp that this light supports."""
+        return self._min_mireds
+
+    @property
+    def max_mireds(self) -> Optional[int]:
+        """Return the warmest color_temp that this light supports."""
+        return self._max_mireds
+
+    @property
+    def white_value(self) -> Optional[int]:
+        """Return the white value of this light between 0..255."""
+        return self._white_value
+
+    @property
+    def effect_list(self) -> Optional[List[str]]:
+        """Return the list of supported effects."""
+        return self._effect_list
+
+    @property
+    def effect(self) -> Optional[str]:
+        """Return the current effect."""
+        return self._effect
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed for a group light."""
+        return False
+
+    async def _async_send_message(self, service, **kwargs):
+        """Send a message to all entities in the group."""
+        for entity_id in self._entity_ids:
+            payload = deepcopy(kwargs)
+            payload[ATTR_ENTITY_ID] = entity_id
+            await self.hass.services.async_call(light.DOMAIN, service, payload)
```

Comment:
Should be blocking=True

Reply:
Why should we be blocking? I mean one potential problem would be that maybe some component takes its time sending the `turn_*` command. Then we shouldn't need to wait for the it before sending the payload to the others, as that would create a weird staggered light on effect if I'm correct.

Reply:
If I would interact with the normal device, it would take the time to finish the command. If I interact with it via proxy, it should take that time to finish the command.

Also, Home Assistant expects commands to be done when the function is done calling. That's why for polling devices it knows it can call `update` right away.

Reply:
How about storing every service call as a task (list) and the use `asyncio.gather`?

Reply:
ehr, you can do it with 1 service call. Just pass the list of entities as ATTR_ENTITY_ID and the light component will figure it all out.

Reply:
> Just pass the list of entities as ATTR_ENTITY_ID and the light component will figure it all out.

Oh, that's certainly good to know, will do.

> Problem with light.async_turn_on is that it's non blocking (see other comment)
> Also, are all kwargs for turn_on and turn_off the same as service attributes? 

Ok, so 1. we need to be able to do the service call with blocking mode + 2. we shouldn't simply pass the kwargs to the service call, as that could break in the future (bc we want to maximize maintainability). 

So ideally I think we would be using something like [`light.async_turn_on`](https://github.com/home-assistant/home-assistant/blob/e82b358831ff1e2b5f631f724b47edfc2e14ec4b/homeassistant/components/light/__init__.py#L163-L188) for the 2. problem, but that doesn't support blocking mode. In my view there are 2 possibilities now (of which I prefer the former):

* Extend `light.async_turn_on` to accept `blocking` kwarg.
  * Upside: we have 1 place where the kwargs -> service call logic sits (DRY... 🤓)
  * Downside: we'd need to add the optional argument + probably also return the task object
  * Or we create a *coroutine* similar to `async_turn_on` that is called from the existing `async_turn_on` and that provides a blocking arg.
* We "bodge" the kwarg -> service call logic into the group platform.
  * Downside: not nice 😥

Reply:
I guess it's nicer to just keep this logic inside the group platform. It's the platforms responsibility to keep working.

Otherwise we add the constraint that `light.async_turn_on` has the same args as `Entity.turn_on`.

Reply:
👍